vortex-array/src/array/bool/compute/take.rs

```diff
@@ -1,55 +1,91 @@
 use arrow_buffer::BooleanBuffer;
 use itertools::Itertools;
 use num_traits::AsPrimitive;
-use vortex_dtype::match_each_integer_ptype;
+use vortex_dtype::{match_each_integer_ptype, NativePType};
 use vortex_error::VortexResult;
+use vortex_mask::Mask;
+use vortex_scalar::Scalar;
 
-use crate::array::{BoolArray, BoolEncoding};
-use crate::compute::TakeFn;
+use crate::array::{BoolArray, BoolEncoding, ConstantArray, PrimitiveArray};
+use crate::compute::{fill_null, TakeFn};
 use crate::variants::PrimitiveArrayTrait;
 use crate::{Array, IntoArray, IntoArrayVariant};
 
 impl TakeFn<BoolArray> for BoolEncoding {
     fn take(&self, array: &BoolArray, indices: &Array) -> VortexResult<Array> {
-        let validity = array.validity();
-        let indices = indices.clone().into_primitive()?;
-
-        // For boolean arrays that roughly fit into a single page (at least, on Linux), it's worth
-        // the overhead to convert to a Vec<bool>.
-        let buffer = if array.len() <= 4096 {
-            let bools = array.boolean_buffer().into_iter().collect_vec();
-            match_each_integer_ptype!(indices.ptype(), |$I| {
-                take_byte_bool(bools, indices.as_slice::<$I>())
-            })
-        } else {
-            match_each_integer_ptype!(indices.ptype(), |$I| {
-                take_bool(&array.boolean_buffer(), indices.as_slice::<$I>())
-            })
+        let indices_nulls_zeroed = match indices.validity_mask()? {
+            Mask::AllTrue(_) => indices.clone(),
+            Mask::AllFalse(_) => {
+                return Ok(
+                    ConstantArray::new(Scalar::null_typed::<bool>(), indices.len()).into_array(),
```

Comment:
Is this scalar definitely the right type? Seems easy to get the wrong nullability